apache/netbeans
# [NETBEANS-4946] - Improve JAVA/JAKARTA EE with Ant based projects

Commit 1. Improve JakartaEE8/JAVAEE8 support with Ant

- Add support for app-client version 8 when using ant
- Add methods that detect support for CDI 1.0 and CDI 2.0
- Add missing resource and license file
- Add missing deployment descriptor for JEE 8
- Remove use of deprecated method "getServerID" in one class
- Remove unnecessary validation that checks if running JDK 7
- Remove use of deprecated method ".getDefault().getJ2eePlatform"
- Use try-with-resources
- Fix memory leak by closing the PrintWriter
- Add missing validations for JEE7 and higher
- Remove unnecessary use of Integer constructor
- Fix wrong way of comparing Integers
- Remove unnecessary assignment inside an if condition
- Add missing validations for EAR applications
- Wrong profile for JAKARTA_EE_8_FULL
- Add missing version for JEE 8
- Add missing validation for JEE8 Application
- Add missing validation for EJB 3.2
- Add validation for WebFragment 3.1
- Add missing MIME_PREFIX for JEE 8
- Add missing validations for JakartaEE8
- Change logic to define isEE6Plus variable
- New logic to evaluate when J2EE is used
- New logic to evaluate the profile to use
- Change mdd files to implement and extend local classes
- Add 1 TODO (make work web-fragment version 4.0)

Commit 2. Update javac.source and compilerargs properties in some Java EE modules
Commit 3. Cleanup

- Add @Override tags and other small fixes

Commit 4. Add logging for test purposes (I will remove added logging)
Commit 5. Add PrintWriter to the try-with-resource construct
Commit 6. Remove unnecessary parameter
Commit 7. Add curly braces for better readability
Commit 8. Fix wrong specification version
Commit 9. Add missing validations for JDK 11, 8 and 7 when setting the target source level
Commit 10. Remove redundant word in javadoc
Commit 11. Add curly braces and align code for better readability
Commit 12. Add logging for testing purposes (I will remove added logging)

Testing:
- Full build done
- Verify successful execution of unit tests for 14 modules [`j2ee.clientproject`-`j2ee.kit`] and `web.beans`
- Verify successful execution of `ant verify-libs-and-licenses`
- Tested with Glassfish 5.1 and TomEE 8
- Create EAR, EJB, WEB and Application Client with JEE 5, 6, 6 w/CDI , 7 , 8 and JAKARTA EE 8 
- Verify the correct creation of deployment descriptors and CDI files for every profile in the projects mention above
- Verify that every project can clean and build successfully
![NETBEANS-4946-1](https://user-images.githubusercontent.com/9832133/104077746-f21fee80-5212-11eb-8d42-db02a0d7badf.png)

Things to improve:
- Could not make work web-fragment version 4.0, it was not working from the beggining. Left the code commented.

Doubts:
- Don't know where to obtain or to create correct mdd files that will work correctly with NetBeans, I modify the `mdd` files that generate the models for JEE 8 Application and Application-client classes. I tried to do the same with web-fragment 4.0 but it failed to compile.

I left some logging for testing/review purposes, I will remove them and squash the commits at the end.
